public/app/features/explore/utils/panelsRegistry.tsx

```diff
@@ -0,0 +1,227 @@
+  KnownVisualizationType,
+  LoadingState,
+  PanelProps,
+  ScopedVars,
+} from '@grafana/data';
+import { selectors } from '@grafana/e2e-selectors';
+import { Collapse, useTheme2 } from '@grafana/ui';
+// // TODO: probably needs to be exported from ui directly
+import { FILTER_FOR_OPERATOR, FILTER_OUT_OPERATOR, FilterItem } from '@grafana/ui/src/components/Table/types';
+import { getAllPanelPluginMeta } from 'app/features/panel/state/util';
+import { importPanelPlugin } from 'app/features/plugins/importPanelPlugin';
+
+import { ExploreGraph } from '../ExploreGraph';
+import { ExploreGraphLabel } from '../ExploreGraphLabel';
+import LogsContainer from '../LogsContainer';
+import { NodeGraphContainer } from '../NodeGraphContainer';
+import TableContainer from '../TableContainer';
+
+/**
+ * Based on the visualizationType try to find a specific panel instance. There are bunch of hardcoded ones but in the
+ * end this will search all the installed panel and check their visualizationType to match. At this moment there
+ * isn't any defined override behaviour so it's not possible to override for example table panel with custom ones
+ * but this allows selecting 3rd party panels.
+ * @param visualizationType
+ */
+export async function getPanelForVisType(visualizationType: string): Promise<React.ComponentType<ExplorePanelProps>> {
+  switch (visualizationType) {
+    case 'graph': {
+      return GraphPanel;
+    }
+    case 'table': {
+      return TablePanel;
+    }
+
+    case 'nodeGraph': {
+      return NodeGraphPanel;
+    }
+
+    case 'logs': {
+      return LogsPanel;
+    }
+
+    default: {
+      const panels = getAllPanelPluginMeta();
+      for (const panel of panels) {
+        if (panel.visualizationType?.includes(visualizationType)) {
+          const panelPlugin = await importPanelPlugin(panel.id);
+          // If there is explorePanel component use that.
+          if (panelPlugin.explorePanel) {
+            return panelPlugin.explorePanel;
+          } else if (panelPlugin.panel) {
+            return makePanelExploreCompatible(panelPlugin.panel!);
+          } else {
+            // Not sure if this can reasonably happen. If this error out we probably catch this at a panel boundary.
+            throw new Error(`Panel plugin does not define any panel component. panel.id=${panel.id}`);
+          }
+        }
+      }
+      // Probably ok fallback but maybe it makes sense to throw or show some info message that we did not find anything
+      // better.
+      return TablePanel;
+    }
+  }
+}
+
+/**
+ * Wrap panel adding a transform so we can use existing dashboard panels in Explore without modification.
+ * @param Panel
+ */
+function makePanelExploreCompatible(Panel: React.ComponentType<PanelProps>): React.ComponentType<ExplorePanelProps> {
+  return function CompatibilityWrapper(props: ExplorePanelProps) {
+    // This transform may not be 100% perfect so we may need to use some sensible zero/empty/noop values. We will have
```

Comment:
1) I'm not sure we want to have this functionality as it has some implications. Explore uses custom visualizations because it supports flows that are not possible to do in Dashboards. That's why Explore needs different panels. Reusing regular/dashboard panels sounds cool, but this will require looking more into customization/settings, maybe changing how "Explore to Dashboard" works. I don't think this direction is the current priority especially as you mentioned _If there are bugs we will have to fix them somehow._ We don't want this to slow down other projects.

2) Having said that I'm wondering if it'd make more sense to have `ExplorePanelPlugin` as a separate plugin type. At the moment API allows creating additional Explore panel kind of attached to regular panels, but what about Explore-only panels? 

3) Could the logic to match the visualization type reuse existing suggestions API? If we had Explore-only plugins suggestions could simply read preferredVisualizationType in runtime to give suggestions (Explore will pick the best which at this moment will be only one). It's a bit confusing that there's `visualizationType` in plugin's meta but it's used only in Explore while Dashboards use suggestions API.

Reply:
This was discussed quite a bit before you got back so I will try to summarize:
1.This seemed to be a big point that we actually want this to work sort of out of the box with minimal changes. Ie the idea is that you could use visualizations as they are and enhance them later. For example the time series panel did not have much of an Explore specific features until very recently. Or the changes may be very minimal (like with logs where for long time only difference was the ability to expand a row). So I don't think it required for explore visualizations to be different, it's more nice to have and you can start with something more simple. "Explore to Dashboard" probably needs some changes but IMHO, even without it, it just won't suggest the same thing and can still be used no problem.

2. Don't think anybody wanted Explore only panels. We did that only once briefly with traces and we ended up having them in dashboards anyway in the end. So I don't think there is an important usecase for that. If the panel isn't useful in dashboard people won't use it but there isn't that much additional work for the developer so we would not gain much by having Explore only panels.
3. Merging these somehow makes sense and I think it will be easier after this change but seems like too many things at once to put it into single PR. The suggestion API is working from the shape of the data which could be used as fallback to preferredVisualizationType but they are not exclusive I think.

Reply:
Thanks for explanation 👏 

1. (...) So I don't think it required for explore visualizations to be different (...)

There's a couple of use cases like _adding a label filter to a log query by clicking on the label_, _logs pagination_, _logs settings_, _graph panel has additional visualization type selector_, _focusing on spanId in trace view_, _data links won't work in Dashboards context_, etc. But if I understand correctly, your point here is that we should be able to use **any** panel in Explore and additonally have tweaked versions provided with `setExplorePanel`, right? 

Do you think the experimental `setExplorePanel(...)` would become part of public API as it is or in your plan it's just an intermediate step? It looks quite confusing to me. There are methods and props that relate to `panel` property (like `defaults`, `noPadding`, etc.) that I think may not be applicable to `explorePanel`. At the same time there's `visualizationType` that applies to `explorePanel` only. Just wondering but would it make sense to have `setExplorePanelPlugin(...)` and expose a nested plugin for Explore?